lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -1108,18 +924,51 @@ def apply_label_matchers(self, rules):
                     if label in rule["labels"]:
                         topology[label] = rule["labels"][label]
 
-                rule["expr"] = self._apply_label_matcher(rule["expr"], topology)
+                rule["expr"] = self.inject_label_matchers(rule["expr"], topology)
         return rules
 
-    def _apply_label_matcher(self, expression, topology):
+    def validate_alert_rules(self, rules: dict) -> Tuple[bool, str]:
+        """Will validate correctness alert rules, returning a boolean and any errors."""
+        if not self.path:
+            logger.debug("`cos-tool` unavailable. Not validating alert correctness.")
+            return True, ""
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            rule_path = Path(tmpdir + "/validate_rule.yaml")
+
+            # Smash "our" rules format into what upstream actually uses, which is more like:
+            #
+            # groups:
+            #   - name: foo
+            #     rules:
+            #       - alert: BlahBlah
+            #         expr: up
+            #       - alert: Haha
+            #         expr: up
+            transformed_rules = {"groups": []}  # type: ignore
+            for rule in rules["groups"]:
+                transformed = {"name": str(uuid.uuid4()), "rules": [rule]}
+                transformed_rules["groups"].append(transformed)
+
+            rule_path.write_text(yaml.dump(rules))
```

Comment:
Is `tempfile.TemporaryDirectory()` guaranteed to succeed in the k8s context?
Is there an advantage in communicating this over stdin?

Reply:
Yes, pods can still make tempdirs.

The advantage is that the rule parsing code in Go, from upstream, expects to find a file. We _could_ pass stdin and read in `[]byte` in Go, but that means that it would not be possible to validate multiple files in a single command if we ever need to, and this is a use case supported by `cortex-tool` and `mimir-tool`, so matching behavior there makes sense.